File: detection/anomaly.py
from __future__ import annotations
import logging
import math
import queue
import threading
from collections import defaultdict, deque
from dataclasses import dataclass
from datetime import datetime
from typing import Deque, Dict, Set, Tuple

from parser.packet import Packet

logger = logging.getLogger(__name__)


@dataclass
class AnomalyAlert:
    timestamp: datetime
    type: str
    description: str
    value: float = 0.0
    mean: float = 0.0
    std_dev: float = 0.0
    z_score: float = 0.0
    src_ip: str = ""


class Detector:
    def __init__(self, window_size: int = 100, threshold: float = 3.0):
        self.window_size = window_size
        self.threshold = threshold
        self._window: Deque[float] = deque(maxlen=window_size)
        self._lock = threading.Lock()
        self.alert_queue: queue.Queue[AnomalyAlert] = queue.Queue(maxsize=200)

        # Port scan tracking: src_ip -> {(dst_port, timestamp)}
        self._src_ports: Dict[str, Dict[int, datetime]] = defaultdict(dict)

        self._current_tick: int = 0
        self._current_count: float = 0.0

    def observe(self, p: Packet) -> None:
        with self._lock:
            tick = int(p.timestamp.timestamp())

            if tick != self._current_tick:
                if self._current_tick != 0:
                    self._window.append(self._current_count)
                    self._check_rate(p.timestamp)
                self._current_tick = tick
                self._current_count = 0.0

            self._current_count += 1

            if p.protocol.upper() == "TCP" and p.dst_port:
                src = p.src_ip
                self._src_ports[src][p.dst_port] = p.timestamp

                # Expire entries older than 10 seconds
                cutoff = p.timestamp.timestamp() - 10
                self._src_ports[src] = {
                    port: ts for port, ts in self._src_ports[src].items()
                    if ts.timestamp() >= cutoff
                }

                if len(self._src_ports[src]) >= 20:
                    self._emit(AnomalyAlert(
                        timestamp=p.timestamp,
                        type="port_scan",
                        description=f"Port scan: {src} contacted 20+ distinct ports in 10s",
                        value=float(len(self._src_ports[src])),
                        src_ip=src,
                    ))
                    self._src_ports[src] = {}

    def _check_rate(self, ts: datetime) -> None:
        if len(self._window) < 10:
            return
        data = list(self._window)
        mean, stddev = _stats(data)
        if stddev == 0:
            return
        latest = data[-1]
        z = (latest - mean) / stddev
        if abs(z) >= self.threshold:
            logger.warning("Packet rate anomaly: count=%.0f mean=%.1f stddev=%.1f z=%.2f",
                           latest, mean, stddev, z)
            self._emit(AnomalyAlert(
                timestamp=ts,
                type="packet_rate",
                description="Packet rate anomaly detected",
                value=latest,
                mean=mean,
                std_dev=stddev,
                z_score=z,
            ))

    def _emit(self, a: AnomalyAlert) -> None:
        try:
            self.alert_queue.put_nowait(a)
        except queue.Full:
            pass


def _stats(data: list) -> Tuple[float, float]:
    n = len(data)
    if n == 0:
        return 0.0, 0.0
    mean = sum(data) / n
    variance = sum((x - mean) ** 2 for x in data) / n
    return mean, math.sqrt(variance)

File: detection/anomaly.go
package detection

import (
	"log"
	"math"
	"sync"
	"time"

	"github.com/Alexmaster12345/IntrusionShield/parser"
)

// AnomalyAlert describes a statistical anomaly event.
type AnomalyAlert struct {
	Timestamp   time.Time
	Type        string  // "packet_rate", "port_scan", "dns_flood"
	Description string
	Value       float64
	Mean        float64
	StdDev      float64
	ZScore      float64
	SrcIP       string
}

// Detector tracks per-second packet rates and raises anomaly alerts
// when the Z-score exceeds the configured threshold.
type Detector struct {
	windowSize int
	threshold  float64
	window     []float64 // rolling packet-count samples
	mu         sync.Mutex

	// Per-source connection tracking for port scan detection
	srcPorts map[string]map[uint16]time.Time

	AlertCh chan AnomalyAlert

	// Current-second counters
	currentTick  int64
	currentCount float64
}

// NewDetector creates an anomaly detector with the given window size and Z-score threshold.
func NewDetector(windowSize int, threshold float64) *Detector {
	d := &Detector{
		windowSize: windowSize,
		threshold:  threshold,
		window:     make([]float64, 0, windowSize),
		srcPorts:   make(map[string]map[uint16]time.Time),
		AlertCh:    make(chan AnomalyAlert, 200),
	}
	return d
}

// Observe feeds a packet into the anomaly detector.
func (d *Detector) Observe(p *parser.Packet) {
	d.mu.Lock()
	defer d.mu.Unlock()

	tick := p.Timestamp.Unix()

	if tick != d.currentTick {
		// New second — flush the previous second's count into the window
		if d.currentTick != 0 {
			d.addSample(d.currentCount)
			d.checkRateAnomaly(p.Timestamp)
		}
		d.currentTick = tick
		d.currentCount = 0
	}
	d.currentCount++

	// Port scan detection: track unique dst ports per src IP in 10-second window
	if p.Protocol == "TCP" && p.DstPort > 0 {
		src := p.SrcIP.String()
		if d.srcPorts[src] == nil {
			d.srcPorts[src] = make(map[uint16]time.Time)
		}
		d.srcPorts[src][p.DstPort] = p.Timestamp

		// Expire old entries and count unique ports in last 10s
		cutoff := p.Timestamp.Add(-10 * time.Second)
		for port, ts := range d.srcPorts[src] {
			if ts.Before(cutoff) {
				delete(d.srcPorts[src], port)
			}
		}
		if len(d.srcPorts[src]) >= 20 {
			d.emit(AnomalyAlert{
				Timestamp:   p.Timestamp,
				Type:        "port_scan",
				Description: "Port scan: " + src + " contacted 20+ distinct ports in 10s",
				Value:       float64(len(d.srcPorts[src])),
				SrcIP:       src,
			})
			// Reset to avoid flooding
			d.srcPorts[src] = make(map[uint16]time.Time)
		}
	}
}

func (d *Detector) addSample(v float64) {
	if len(d.window) >= d.windowSize {
		d.window = d.window[1:]
	}
	d.window = append(d.window, v)
}

func (d *Detector) checkRateAnomaly(ts time.Time) {
	if len(d.window) < 10 {
		return
	}
	mean, stddev := stats(d.window)
	if stddev == 0 {
		return
	}
	latest := d.window[len(d.window)-1]
	z := (latest - mean) / stddev

	if math.Abs(z) >= d.threshold {
		log.Printf("[anomaly] packet rate spike: count=%.0f mean=%.1f stddev=%.1f z=%.2f", latest, mean, stddev, z)
		d.emit(AnomalyAlert{
			Timestamp:   ts,
			Type:        "packet_rate",
			Description: "Packet rate anomaly detected",
			Value:       latest,
			Mean:        mean,
			StdDev:      stddev,
			ZScore:      z,
		})
	}
}

func (d *Detector) emit(a AnomalyAlert) {
	select {
	case d.AlertCh <- a:
	default:
	}
}

func stats(data []float64) (mean, stddev float64) {
	if len(data) == 0 {
		return
	}
	sum := 0.0
	for _, v := range data {
		sum += v
	}
	mean = sum / float64(len(data))

	variance := 0.0
	for _, v := range data {
		d := v - mean
		variance += d * d
	}
	stddev = math.Sqrt(variance / float64(len(data)))
	return
}
